### PR TITLE
Add GaussianRandomProjector and SparseRandomProjector

### PR DESCRIPTION
**Pull request by MaxHalford**
Same as scikit-learn's `random_projection` module. This is yet another way to deal with inputs that have many features.

### CODE REVIEW

File: river/preprocessing/random_projection.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+from river import base
+
+
+class GaussianRandomProjector(base.Transformer):
+    """Gaussian random projector.
+
+    This transformer reduces the dimensionality of inputs through Gaussian random projection.
+
+    The components of the random projections matrix are drawn from `N(0, 1 / n_components)`.
```
Comment (smastelini) [inline river/preprocessing/random_projection.py:13]: Good stuff!! 

Is there a missing `** 0.5` in the docstring?

File: river/preprocessing/test_random_projection.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+def test_sparse_random_projector_dot_product():
+    dataset = datasets.TrumpApproval()
+    projector = preprocessing.SparseRandomProjector(n_components=3, density=0.5)
+
+    for x, y in dataset:
+        y = projector.transform_one(x)
+        y_arr = np.array(list(y.values()))
+        x_arr = np.array(list(x.values()))
+        P = np.array(
```
Comment (smastelini) [inline river/preprocessing/test_random_projection.py:28]: Could you add a comment here to clarify the meaning of this operation for the future generations? XD

**Review by smastelini — APPROVED** [review]
LGTM :)

File: river/preprocessing/test_random_projection.py (hunk 2 above)
Comment (MaxHalford) [reply river/preprocessing/test_random_projection.py:28]: Good call!

File: river/preprocessing/random_projection.py (hunk 1 above)
Comment (MaxHalford) [reply river/preprocessing/random_projection.py:13]: Good catch :)

**Pull request merged**